src/hotspot/share/runtime/javaThread.inline.hpp

```diff
@@ -196,6 +196,11 @@ void JavaThread::enter_critical() {
   _jni_active_critical++;
 }
 
+inline void JavaThread::set_is_attaching_via_jni() {
+  _jni_attach_state = _attaching_via_jni;
+  OrderAccess::fence();
```

Comment:
No need for a fence here as this should be set before the thread has been "published" and is visible to anyone else. In contrast `set_done_attaching_via_jni` is set after the thread has been published.

Reply:
I had the vague notion of not wanting to tie `set_done_attaching_via_jni` to an exact point in time. But that was not well thought out anyway. With a factory method, this problem disappears.